From 2a66f42e57310eece52b626c4b0f14476bae8a20 Mon Sep 17 00:00:00 2001
From: Bin Han <bin.b.han@intel.com>
Date: Tue, 13 May 2014 18:26:49 +0800
Subject: [PATCH 087/117] v4l2: copy reserved and reserved2 of v4l2_buffer in
 compat_ioctl32

In the per-frame parameter setting feature, camera HAL needs to use the
reserved2 in v4l2_buffer to specify which parameter it intends to apply and
also use reserved2 to indicate if this frame buffer do not need a per-frame
parameter. So v4l2 needs to copy the reserved2 and reserved in compat_ioctl32
for this purpose.

(Note that this is a workaround due to the atomisp driver/userspace
storing data in those reserved fields which are supposed to be
ignored.  Needs a long term fix.)

Change-Id: I0f6d26f4ccf05931324f9ba73be6e2c12f48a2f9
Signed-off-by: Bin Han <bin.b.han@intel.com>
---
 drivers/media/v4l2-core/v4l2-compat-ioctl32.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/media/v4l2-core/v4l2-compat-ioctl32.c b/drivers/media/v4l2-core/v4l2-compat-ioctl32.c
index a191f27..bcd53ee 100644
--- a/drivers/media/v4l2-core/v4l2-compat-ioctl32.c
+++ b/drivers/media/v4l2-core/v4l2-compat-ioctl32.c
@@ -392,7 +392,9 @@ static int get_v4l2_buffer32(struct v4l2_buffer *kp, struct v4l2_buffer32 __user
 		get_user(kp->index, &up->index) ||
 		get_user(kp->type, &up->type) ||
 		get_user(kp->flags, &up->flags) ||
-		get_user(kp->memory, &up->memory))
+		get_user(kp->memory, &up->memory) ||
+		get_user(kp->reserved2, &up->reserved2) ||
+		get_user(kp->reserved, &up->reserved))
 			return -EFAULT;
 
 	if (V4L2_TYPE_IS_OUTPUT(kp->type))
-- 
1.9.3

